Building with -DBUILD_SHARED_LIBS=ON on macOS fails with linking errors

Hi

I'm building llvm-mos/llvm-mos on macOS 12.1 Monterey with the command-lines

```
$ cmake   -C clang/cmake/caches/MOS.cmake -G Ninja -DBUILD_SHARED_LIBS=ON -DLLVM_PARALLEL_COMPILE_JOBS=8 -DLLVM_PARALLEL_LINK_JOBS=2 -S llvm -B build
$ cmake -B build --parallel 8

```

The build gets quite far, but then fails with a slew of linking errors that seem to indicate that the contents of llvm/lib/Targets/MOS/ have not been included in any of the libraries. I had a quick look through the CMake* files, but didn't see anything obvious.

Setting `-DBUILD_SHARED_LIBS=OFF` works as expected. I prefer the shared library version of the build for rapid development as the binaries are smaller and the build is quicker, but static linking is better for tricky debugging, and I'll be using it for now.

Thank you,

M

PS: I'm going to be using this excellent work of yours to create a fork, privately at first. I'm an enthusiast of the MC6809 processor, and I'm planning to base an MC6809/HD6309 backend on your work. I had a backend assembler/disassembler working a couple of years back  but your work is much more mature, with working CodeGen, relocation and linking. Bravo!

I reproduce this at 9a8981cfd5c72c2e60151ce670cd56acc5b57148 on macOS Ventura 13.0:

    fox@wiesmac build % c++ --version
    Apple clang version 14.0.0 (clang-1400.0.29.102)
    Target: arm64-apple-darwin22.1.0
    Thread model: posix
    InstalledDir: /Library/Developer/CommandLineTools/usr/bin
    fox@wiesmac build % cmake --version
    cmake version 3.24.2

    CMake suite maintained and supported by Kitware (kitware.com/cmake).
    fox@wiesmac build % cmake -B build --parallel=8
    CMake Error: Unknown argument --parallel=8
    CMake Error: Run 'cmake --help' for all supported options.
    fox@wiesmac build % ninja
    [1/2301] Linking CXX shared library lib/libLLVMMOSDisassembler.dylib
    FAILED: lib/libLLVMMOSDisassembler.dylib
    : && /Library/Developer/CommandLineTools/usr/bin/c++ -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -fdiagnostics-color -Os -DNDEBUG -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX12.3.sdk -dynamiclib -Wl,-headerpad_max_install_names -Wl,-dead_strip -o lib/libLLVMMOSDisassembler.dylib -install_name @rpath/libLLVMMOSDisassembler.dylib lib/Target/MOS/Disassembler/CMakeFiles/LLVMMOSDisassembler.dir/MOSDisassembler.cpp.o  -Wl,-rpath,@loader_path/../lib  lib/libLLVMMCDisassembler.dylib  lib/libLLVMMOSInfo.dylib  lib/libLLVMSupport.dylib && :
    Undefined symbols for architecture arm64:
      "llvm::MCConstantExpr::create(long long, llvm::MCContext&, bool, unsigned int)", referenced from:
          (anonymous namespace)::MOSDisassembler::getInstruction(llvm::MCInst&, unsigned long long&, llvm::ArrayRef<unsigned char>, unsigned long long, llvm::raw_ostream&) const in MOSDisassembler.cpp.o
      "llvm::MOSMCExpr::create(llvm::MOSMCExpr::VariantKind, llvm::MCExpr const*, bool, llvm::MCContext&)", referenced from:
          (anonymous namespace)::MOSDisassembler::getInstruction(llvm::MCInst&, unsigned long long&, llvm::ArrayRef<unsigned char>, unsigned long long, llvm::raw_ostream&) const in MOSDisassembler.cpp.o
    ld: symbol(s) not found for architecture arm64
    clang: error: linker command failed with exit code 1 (use -v to see invocation)
    [10/2301] Building CXX object lib/ProfileData/CMakeFiles/LLVMProfileData.dir/SampleProfReader.cpp.o
    ninja: build stopped: subcommand failed.

Fix: #240

--- a/llvm/lib/Target/MOS/CMakeLists.txt
+++ b/llvm/lib/Target/MOS/CMakeLists.txt
@@ -53,17 +53,24 @@ add_llvm_target(MOSCodeGen
   intrinsics_gen
 
   LINK_COMPONENTS
+  Analysis
   AsmPrinter
+  BinaryFormat
   CodeGen
   Core
   GlobalISel
-  MC
   IPO
+  InstCombine
   LTO
+  MC
+  MOSDesc
+  MOSInfo
   Passes
+  Scalar
   SelectionDAG
   Support
   Target
+  TransformUtils
 
   ADD_TO_COMPONENT
   MOS

--- a/llvm/lib/Target/MOS/Disassembler/CMakeLists.txt
+++ b/llvm/lib/Target/MOS/Disassembler/CMakeLists.txt
@@ -2,7 +2,9 @@ add_llvm_component_library(LLVMMOSDisassembler
   MOSDisassembler.cpp
 
   LINK_COMPONENTS
+  MC
   MCDisassembler
+  MOSDesc
   MOSInfo
   Support
 

